### ISSUE TITLE
overlay2: docker push creating overlapped ovls, rendering failures on newer kernels

### ISSUE THREAD

**Issue by omegacoleman** [moby/moby#39663]
**Description**

In short, the overlay subsystem in kernel forbids overlapped layers -- that means, one layer being the upper dir & the lower at the same time. Which is just what `docker push` does.

This is an UB, undefined behavior in before, and with this patch https://github.com/torvalds/linux/commit/146d62e5a5867fbf84490d82455718bfb10fe824 , this action may cause error.

I have not looked into the issue, but I think it's ilogic that `docker push` creates overlapped layers, and moreover, why did it use ovl in the first place?

**Steps to reproduce the issue:**
1. Install a newer kernel (containing that patch), for me I use fedora 30.

```
[root@localhost ~]# uname -a
Linux localhost.localdomain 5.1.20-300.fc30.x86_64 #1 SMP Fri Jul 26 15:03:11 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

2. Pull an image with multiple layers, here I choose mariadb, and tag it so it could be pushed to a personal registry(or any registry not including this image). Notice that the bug happens during push of non-exist image layers, that's why I take the personal registry. 

```
[root@localhost ~]# docker tag mariadb xxx.yyy.zzz/mariadb
```

3. `docker push xxx.yyy.zzz/mariadb`

**Describe the results you received:**

Some of the layers (often the top 4) start to fail & retry

```
[root@localhost ~]# docker push xxx.yyy.zzz/mariadb
The push refers to repository [xxx.yyy.zzz/mariadb]
0681153a129b: Retrying in 1 second 
cd7a8d37f569: Retrying in 2 seconds 
...
```

`journalctl -f -u docker` outputs error messages related to overlay2

```
Aug 04 08:50:05 localhost.localdomain dockerd[3012]: time="2019-08-04T08:50:05.459134956-07:00" level=error msg="Upload failed, retrying: error creating overlay mount to /var/lib/docker/overlay2/aea27291a899b955bbed386a4c4a37a4512586a583c5fa05bc37a82cac229426/merged: device or resource busy"
```

and `dmesg` complains about the overlap:

```
[ 1348.372328] overlayfs: lowerdir is in-use as upperdir/workdir
[ 1394.161732] overlayfs: lowerdir is in-use as upperdir/workdir
```

**Describe the results you expected:**

The push shall be okay.

**Additional information you deem important (e.g. issue happens only occasionally):**

If tried many times running the same push command, the push might be completed. But this issue is first seen on my CI system, and it could get the system unstable.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:20:55 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:19:28 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 19.03.1
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.1.20-300.fc30.x86_64
 Operating System: Fedora 30 (Server Edition)
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 1.92GiB
 Name: localhost.localdomain
 ID: LGLA:TGK2:YCZC:NSHI:G4AU:CI3M:RWQX:2JN4:4AVD:YORS:RIW7:C2AE
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

The above output is gathered from an repr image inside VMWare, but the issue happens on physical machine in the first.

**Comment by omegacoleman** [comment]
@thaJeztah pls take a look at it. I compiled a custom kernel including https://github.com/cilium/packer-ci-build/commit/babb646d5fa075f9042224a2d8548d4ce9e34d32 this patch which solved the issue. But there's obviously sth wrong with `docker push`

**Comment by thaJeztah** [comment]
I'd have to dig into the logic; from a naive thinking; I wonder if the reason is that it's creating the tar for multiple layers in parallel, so on an image with 3 layers (A, B, C)

- Mount A -> create archive for layer A
- Mount A (lower), B (upper) -> diff A..B -> create archive for layer B
- Mount A (lower), B (lower), C (upper) -> diff (A+B)..C -> create archive for layer C

**Comment by thaJeztah** [comment]
ping @dmcgowan @kolyshkin PTAL

**Comment by thaJeztah** [comment]
FWIW, containerd's storage wouldn't have that problem as it (by default) preserves the compressed layers (whereas dockerd currently has to reconstruct the compressed layers when pushing)

**Comment by omegacoleman** [comment]
> I'd have to dig into the logic; from a naive thinking; I wonder if the reason is that it's creating the tar for multiple layers in parallel, so on an image with 3 layers (A, B, C)
> 
> * Mount A -> create archive for layer A
> * Mount A (lower), B (upper) -> diff A..B -> create archive for layer B
> * Mount A (lower), B (lower), C (upper) -> diff (A+B)..C -> create archive for layer C

I may look into it this weekend, but allow me to ask ahead: what prevents us from creating the archive of layer B & C directly from the folder, instead of mounting it first?

**Comment by omegacoleman** [comment]
> FWIW, containerd's storage wouldn't have that problem as it (by default) preserves the compressed layers (whereas dockerd currently has to reconstruct the compressed layers when pushing)

Yes, I covered most of the functions, `docker push` is the only one breaking.

**Comment by fuweid** [comment]
I think it is the same issue to https://github.com/moby/moby/issues/34672

**Comment by fuweid** [comment]
Actually, it seems that containerd has the same issue when the user try to commit a container.  The commit will use the same upperdir to copy all the content into other snapshot. 

```
kernel: overlayfs: upperdir is in-use by another mount, mount with '-o index=off' to override exclusive upperdir protection.
```

cc @thaJeztah @dmcgowan

**Comment by thaJeztah** [comment]
For committing, yes, that sounds plausible, because then it needs the overlay filesystem to calculate the diff

**Comment by thaJeztah** [comment]
@fuweid thanks for that link btw; through that issue, I also noticed we have this issue opened: https://github.com/moby/moby/issues/39475 which indicates it may be a kernel / overlays issue

I see CoreOS modified their kernel to suppress this warning https://github.com/coreos/linux/pull/346

**Comment by kolyshkin** [comment]
@omegacoleman  Wonder if doing `echo off > /sys/module/overlay/parameters/index` makes any difference?

**Comment by omegacoleman** [comment]
@fuweid @thaJeztah I did not repr #34672 with the environment described above, plus the error message in this issue is "lowerdir is in-use as upperdir/workdir", and I don't think it would appear in 4-series kernels. Plus I have double checked that I have no issue with `docker commit`

@kolyshkin It's already off during the repr.
```
[root@localhost ~] cat /sys/module/overlay/parameters/index
N
```

**Comment by vladimiroff** [comment]
I have a similar issue (same error and same message in dmesg) on 5.2.8 with COPY during 'docker build'.

On my machine setting `/sys/module/overlay/parameters/metacopy` to `N` works around it.

**Comment by cameronwp** [comment]
Seeing the same errors as @omegacoleman. `/sys/module/overlay/parameters/index` does not seem to have an effect. `/sys/module/overlay/parameters/metacopy` is already `N`.
```
$ uname -a
Linux localhost 5.2.13-arch1-1-ARCH #1 SMP PREEMPT Fri Sep 6 17:52:33 UTC 2019 x86_64 GNU/Linux

$ docker version
Client:
 Version:           19.03.2-ce
 API version:       1.40
 Go version:        go1.13
 Git commit:        6a30dfca03
 Built:             Tue Sep  3 23:41:01 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.2-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13
  Git commit:       6a30dfca03
  Built:            Tue Sep  3 23:39:37 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.9.m
  GitCommit:        d50db0a42053864a270f648048f9a8b4f24eced3.m
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

$ docker info
Client:
 Debug Mode: false

Server:
 Containers: 8
  Running: 0
  Paused: 0
  Stopped: 8
 Images: 107
 Server Version: 19.03.2-ce
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: d50db0a42053864a270f648048f9a8b4f24eced3.m
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.2.13-arch1-1-ARCH
 Operating System: Arch Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 31.37GiB
 Name: saturn
 ID: FN6Y:UIKP:CQME:DOS4:WTWA:KPM2:5RRA:FVFV:JRKW:XR5Y:V6ON:U6LF
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by awprice** [comment]
I am experiencing this as well after upgrading to 4.19.56. 

4.19.75 was just released (https://lwn.net/Articles/800247/) and includes a fix for an overlay2 regression:
```
Amir Goldstein (1):
      ovl: fix regression caused by overlapping layers detection
```

I can confirm 4.19.75 fixed the issue for me.

**Comment by vladimiroff** [comment]
The same [fix](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=0be0bfd2de9dfdd2098a9c5b14bdd8f739c9165d) has also being released in 5.3.1 and 5.2.17. I can confirm it working on 5.3.1.

**Comment by omegacoleman** [comment]
confirmed fix on 5.3.1, closing this
